### PR TITLE
Adding the function nanmax and its tests

### LINKED ISSUE #22729: <nanmax>

### PR DESCRIPTION
**Pull request by GaelKBertrand**
# PR Description 

## Related Issue 

Close #22729

## Checklist 

- [ ] Did you add a function? Yes 
- [ ] Did you add the tests? Yes 
- [ ] Did you follow the steps we provided? Yes 

### Socials: 

Discord: beekaykoozie

### CODE REVIEW

**Comment by GaelKBertrand** [comment]
Any idea why there so many failing tests despite writing necessary tests and passing locally?

**Comment by yopknopixx** [comment]
@GaelKBertrand I see only one file changed. I think you may have missed adding the actual function implementation.

**Comment by yopknopixx** [comment]
Hey @GaelKBertrand, before moving forward with this I'd suggest you have a look at the PRs of closed issues [here](https://github.com/unifyai/ivy/issues/3856). That will give you a better understanding of what is expected for this task. Thanks!

**Comment by GaelKBertrand** [comment]
@yopknopixx I added all the necessary additions. Now, there are still new failing tests. I can use a detailed technical review/answer. 
Secondly, 'nanmax' is a function that exists in other frameworks like paddle, mxnet, pandas, and mindspore. Are we required to typically only do front-end implementation for (Jax, Tensorflow, Torch, Numpy) or can we add those ones as well?

File: ivy/functional/ivy/statistical.py [hunk 1]
```diff
@@ -1310,3 +1315,40 @@ def einsum(
...
     return current_backend(operands[0]).einsum(equation, *operands, out=out)
+
+
+@with_unsupported_dtypes({"2.0.1 and below": ("float16",)}, "torch")
+@infer_device
+@infer_dtype
+@handle_array_function
+@outputs_to_ivy_arrays
+@inputs_to_ivy_arrays
+def nanmax(a, axis=None, keepdims=False):
```
Comment (yopknopixx) [inline ivy/functional/ivy/statistical.py:1326]: This should call the backend functions you defined.

**Review by yopknopixx — CHANGES_REQUESTED** [review]
Hey @GaelKBertrand, please excuse the delayed response. I can see under files changed that you have implemented `nanmax` twice in each backend, which is unnecessary. I would suggest you have a look at the relevant documentation before moving forward.

File: ivy/functional/ivy/statistical.py (hunk 1 above)
Comment (GaelKBertrand) [reply ivy/functional/ivy/statistical.py:1326]: Why are you not addressing my questions directly? I have asked from the beginning about this. First, there was to follow a merged PR. I did that exactly and 8 files were changed accordingly. This resulted in the same errors as before that were not explained. While I was waiting for your response, an ivy member on discord advised to include the fron-end implementations for each framework in the backend, which I did and resulted into errors. I would like a technical review of the code and pinpointing exactly the problem. For now, I assume it's about directories and duplicity.

Comment (GaelKBertrand) [reply ivy/functional/ivy/statistical.py:1326]: My questions are specifically about the confusion I had, directly emerging from the documentation. If the questions can't be answered directly, I suggest I will be working on this tomorrow onwards and debug it with others who will be willing to answer questions directly.

**Comment by GaelKBertrand** [comment]
Yes. I m digging in deeper in the deep dive doc. I aim to understand all of it. I will list the questions in a few hours and I think this time I will phrase them better and I will be able to put the code in the right directory. So far I don't think I had any issue with the code.

**Comment by GaelKBertrand** [comment]
@yopknopixx @Aarsh2001 @zaeemansari70 
I removed the last mistake and put all code in their respective dirctories. Summary:
1. Function implementation in backend APIs: Jax, Numpy, Torch, Tensorflow.
2. Function implementation in the Ivy API
3. Tests in the tests folders. 

Questions:

1. Should implementation be done also for paddle and mxnet? In this case, we also have various front-end implementations which have the function nanmax that can be added. 
2. There are still errors that I can't fully comprehend despite the implementation and the scripts being correct. I would like some help on this too.

**Comment by yopknopixx** [comment]
1. Yes, you should implement the backend functions in all the backends. If there are compositional implements for `nanmax` in the front-ends of those frameworks, those should be refactored to use the updated backend implementation.
2. One of the factors contributing to failing tests may be the incorrect implementation in `ivy/functional/ivy/statistical.py` as I've pointed out in the review above. Once all the issues in the code are fixed, I can determine if your PR introduces any new failures.
Let me know if you need any further clarification. Thanks!

**Comment by GaelKBertrand** [comment]
Hi @yopknopixx 

1. This sentence here is not clear: "One of the factors contributing to failing tests MAY BE the incorrect implementation in ivy/functional/ivy/statistical.py as I've pointed out in the review above." As of the last contribution, there was no implementation directly in the directory you mentioned. The contribution was in ivy/experimental/statistical.py if you check well. I need to know exactly what you are calling an incorrect implementation (explain using computer science, not simple english phrases) because it's the first time I'm contributing to Ivy. 
2. This brings the question again? Where exactly these functions are placed? Some are in that directory you specified, some are in the experimental directories despite the documentation saying to place them in the experimental ones only. 
3. Regarding the decorators problem, the comments above were not clear. What does this mean in the context of saying it is a wrong implementation MAYBE : "This should call the backend functions you defined.", "outputs_to_ivy_arrays and inputs_to_ivy_arrays are specific to frontend functions."

**Comment by yopknopixx** [comment]
Hey @GaelKBertrand I'm sorry for the confusion. 
1. To clarify, whatever you have implemented which is currently at `ivy/functional/ivy/experimental/statistical.py` is incorrect. However, it MAY NOT be the only thing that is incorrect (this may include bugs with the testing pipeline or other mistakes in your implementation) given the number of errors. Here is an example of a similar correctly implemented function in `ivy/functional/ivy/experimental/statistical.py`:
```
@handle_exceptions
@handle_backend_invalid
@handle_nestable
@handle_out_argument
@to_native_arrays_and_back ### 3. The data conversion that needs to happen here is ivy.Array -> backend array (tf.Tensor or np.array) depending on the current backend. This decorator takes care of that.
@infer_dtype
@handle_device_shifting
def nanmean(
    a: ivy.Array,
    /,
    *,
    axis: Optional[Union[Tuple[int], int]] = None,
    keepdims: bool = False,
    dtype: Optional[Union[ivy.Dtype, ivy.NativeDtype]] = None,
    out: Optional[ivy.Array] = None,
) -> ivy.Array:
    """
    Compute the mean of all non-NaN elements along the specified dimensions.

    Parameters
    ----------
    a
        Input array.
    axis
        Axis or axes along which the means are computed.
        The default is to compute the mean of the flattened array.
    keepdims
        If this is set to True, the axes which are reduced are left in the result
        as dimensions with size one. With this option, the result will broadcast
        correctly against the original a. If the value is anything but the default,
        then keepdims will be passed through to the mean or sum methods of sub-classes
        of ndarray. If the sub-classes methods does not implement keepdims any
        exceptions will be raised.
    dtype
        The desired data type of returned tensor. Default is None.
    out
        optional output array, for writing the result to.

    Returns
    -------
    ret
        The nanmean of the array elements.

    Functional Examples
    -------------------
    >>> a = ivy.array([[1, ivy.nan], [3, 4]])
    >>> ivy.nanmean(a)
    2.6666666666666665
    >>> ivy.nanmean(a, axis=0)
    ivy.array([2.,  4.])
    """
    return ivy.current_backend(a).nanmean(
        a, axis=axis, keepdims=keepdims, dtype=dtype, out=out ### 3. Here you can see the "nanmean" function calling the backend functions defined for "nanmean"
    )
```
2. The placement of these functions should be in `experimental` directories, which you have done correctly. However, placement is not relevant as there are problems with the implementation (see comments in the above code snippet marked with `###` and the explanation below)
3. The `nanmax` function in `ivy/functional/ivy/experimental/statistical.py` is expected to take `ivy.Array` objects as inputs, convert them to backend-arrays (also called native arrays) and pass them to the current backend implementation for `nanmax`  (This should call the backend functions you defined).
The `outputs_to_ivy_arrays` function is used to convert native array returns to `ivy.Array`s (this is covered along with converting native arrays returned after executing the backend function to `ivy.Array`s by `to_native_arrays_and_back` in the above example) and the `inputs_to_ivy_arrays` is used with front-end functions to convert front-end arrays to `ivy.Array`s which is not relevant here.
I hope this clears up any confusion that you have regarding this task. If not, I encourage you to refer to the documentation and similar previously implemented functions in the Ivy codebase. Please feel free to reach out to me, in case there is still a lack of understanding after making the fixes suggested in the above comments and going through the docs (if necessary).

**Review by yopknopixx — CHANGES_REQUESTED** [review]
Please implement changes as specified in the latest comment.
The files: `combined_failures.txt`, `.idea/vcs.xml`, `.idea/runConfigurations/_template__of_py_test.xml`, `.idea/misc.xml`, `.idea/ivy.iml` should not have any changes. Please fix that.

File: ivy/functional/backends/tensorflow/experimental/statistical.py [hunk 2]
```diff
@@ -781,3 +781,21 @@ def cummin(
         return cummin_x
     else:
         return tf.cast(cummin_x, dtype)
+
+
+# nanmax
+def nanmax(a: tf.Tensor, axis=None, keepdims=False, name=None):
+    if axis is not None:
+        a = tf.transpose(a, perm=axis[::-1])
+        # Move specified axis to the first dimension
```
Comment (rishabgit) [inline ivy/functional/backends/tensorflow/experimental/statistical.py:790]: which code line is this comment for? not sure i follow 😅

File: ivy/functional/ivy/experimental/statistical.py [hunk 3]
```diff
@@ -894,3 +896,52 @@ def cummin(
...
+@ivy.handle_out_argument
+@ivy.to_native_arrays_and_back
+@ivy.handle_device_shifting
+def nanmax(
+    a: ivy.Array,
+    /,
+    *,
+    axis: Optional[Union[Tuple[int], int]] = None,
+    keepdims: bool = False,
+) -> ivy.Array:
```
Comment (rishabgit) [inline ivy/functional/ivy/experimental/statistical.py:915]: missing `out`

File: ivy/functional/ivy/experimental/statistical.py [hunk 4]
```diff
@@ -894,3 +896,52 @@ def cummin(
...
+    array(5., dtype=float32)
+    """
+    # Create a mask for NaN values in the input array
+    mask = ivy.isnan(a)
+
+    # Replace NaN values with negative infinity
+    masked_a = ivy.where(mask, -ivy.inf, a)
+
+    # Compute the maximum value of the masked array
+    backend = ivy.current_backend(a)
```
Comment (rishabgit) [inline ivy/functional/ivy/experimental/statistical.py:946]: backend doesn't need to be instantiated this way, check out other fns for convention

**Review by rishabgit — CHANGES_REQUESTED** [review]
Hi! I've added some comments, feel free to re-request review when done! Thanks 😁

File: ivy/functional/ivy/experimental/statistical.py (hunk 4 above)
Comment (GaelKBertrand) [reply ivy/functional/ivy/experimental/statistical.py:946]: @rishabgit, in local testing what could be the origin of this error? 
[Test Results - pytest_for_test_statistical_test_nanmax.pdf](https://github.com/unifyai/ivy/files/12833139/Test.Results.-.pytest_for_test_statistical_test_nanmax.pdf)

**Pull request closed without merging**